File: tools/opensdk/typescript/transform.ts
// SPDX-License-Identifier: Apache-2.0

type Input = Readonly<{ source: string; prefix: string }>;

let encoded = "";
for await (const chunk of process.stdin) {
  encoded += chunk;
  if (Buffer.byteLength(encoded, "utf8") > 1024 * 1024) {
    throw new Error("input exceeds 1 MiB");
  }
}
const value: unknown = JSON.parse(encoded);
if (!isInput(value)) throw new Error("input must contain exactly source and prefix strings");
if (value.source.length > 64 * 1024 || value.prefix.length > 1024) {
  throw new Error("transform input exceeds field limits");
}
if (!/^[\x00-\x7f]*$/.test(value.source) || !/^[\x00-\x7f]*$/.test(value.prefix)) {
  throw new Error("uppercase-v1 accepts ASCII input only");
}
const transformed = `${value.prefix}${value.source.toUpperCase()}`;
process.stdout.write(`${JSON.stringify({
  schema: "dev.pliegors.build-transform/v1",
  mediaType: "text/plain; charset=utf-8",
  bytesBase64: Buffer.from(transformed, "utf8").toString("base64"),
})}\n`);

function isInput(value: unknown): value is Input {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  return Object.keys(record).sort().join(",") === "prefix,source" &&
    typeof record.source === "string" && typeof record.prefix === "string";
}

File: tools/opensdk/python/transform.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0

import base64
import json
import sys

encoded = sys.stdin.buffer.read(1024 * 1024 + 1)
if len(encoded) > 1024 * 1024:
    raise ValueError("input exceeds 1 MiB")
value = json.loads(encoded)
if (
    not isinstance(value, dict)
    or set(value) != {"source", "prefix"}
    or not isinstance(value["source"], str)
    or not isinstance(value["prefix"], str)
):
    raise ValueError("input must contain exactly source and prefix strings")
if len(value["source"]) > 64 * 1024 or len(value["prefix"]) > 1024:
    raise ValueError("transform input exceeds field limits")
if not value["source"].isascii() or not value["prefix"].isascii():
    raise ValueError("uppercase-v1 accepts ASCII input only")

transformed = f'{value["prefix"]}{value["source"].upper()}'.encode("utf-8")
output = {
    "schema": "dev.pliegors.build-transform/v1",
    "mediaType": "text/plain; charset=utf-8",
    "bytesBase64": base64.b64encode(transformed).decode("ascii"),
}
print(json.dumps(output, ensure_ascii=True, separators=(",", ":")))
